certora/KilnBase.spec

```diff
@@ -0,0 +1,316 @@
+    require(dai == sell());
+
+    uint256 ward = wards(e.msg.sender);
+    uint256 locked = lockedGhost();
+
+    uint256 balanceKiln = dai.balanceOf(currentContract);
+    uint256 balanceDst = dai.balanceOf(dst);
+
+    bool dstSameAsKiln = currentContract == dst;
+
+    rug@withrevert(e, dst);
+
+    bool revert1 = e.msg.value > 0;
+    bool revert2 = ward != 1;
+    bool revert3 = locked != 0;
+    bool revert4 = !dstSameAsKiln && balanceKiln + balanceDst > max_uint256;
+
+    assert(revert1 => lastReverted, "revert1 failed");
+    assert(revert2 => lastReverted, "revert2 failed");
+    assert(revert3 => lastReverted, "revert3 failed");
+    assert(revert4 => lastReverted, "revert4 failed");
+
+    assert(lastReverted => revert1 || revert2 || revert3 ||
+                           revert4, "Revert rules are not covering all the cases");
+}
+
+// Verify correct storage changes for not reverting fire
+rule fire() {
+    env e;
+
+    require(dai == sell());
+    require(token == buy());
+    require(pool == pool());
+
+    uint256 daiBalanceKilnBefore = dai.balanceOf(currentContract);
+    uint256 daiBalancePoolBefore = dai.balanceOf(pool);
+    uint256 tokenBalanceKilnBefore = token.balanceOf(currentContract);
+    uint256 tokenBalancePoolBefore = token.balanceOf(pool);
+    uint256 daiSupplyBefore = dai.totalSupply();
+    uint256 tokenSupplyBefore = token.totalSupply();
+    uint256 lot = lot();
+
+    uint256 minAmt = min(daiBalanceKilnBefore, lot);
+
+    fire(e);
+
+    uint256 daiBalanceKilnAfter = dai.balanceOf(currentContract);
+    uint256 daiBalancePoolAfter = dai.balanceOf(pool);
+    uint256 tokenBalanceKilnAfter = token.balanceOf(currentContract);
+    uint256 tokenBalancePoolAfter = token.balanceOf(pool);
+    uint256 daiSupplyAfter = dai.totalSupply();
+    uint256 tokenSupplyAfter = token.totalSupply();
+    uint256 zzzAfter = zzz();
+
+    assert(daiSupplyAfter == daiSupplyBefore,                          "assert1 failed");
+    assert(daiBalanceKilnAfter == (daiBalanceKilnBefore - minAmt),     "assert2 failed");
+    assert(daiBalancePoolAfter == (daiBalancePoolBefore + minAmt),     "assert3 failed");
+    assert(tokenBalancePoolAfter == (tokenBalancePoolBefore - minAmt), "assert4 failed");
+    assert(tokenSupplyAfter == (tokenSupplyBefore - minAmt),           "assert5 failed");
+    assert(zzzAfter == e.block.timestamp,                              "assert6 failed");
+    assert(tokenBalanceKilnAfter == tokenBalanceKilnBefore,            "assert7 failed");
+}
+
+// Verify revert rules on fire
+rule fire_revert() {
+    env e;
+
+    require(dai == sell());
+    require(token == buy());
+    require(authority == token.authority());
+    require(pool == pool());
+
+    uint256 locked = lockedGhost();
+
+    bool stop = token.stopped();
+    address tokenOwner = token.owner();
+    bool canCall = authority.canCall(e, currentContract, token, 0x42966c6800000000000000000000000000000000000000000000000000000000); // burn(uint256)
```

Comment:
Tbh it seems like overkill to be using a full DSToken for the burn token here--we only care about testing kiln bytecode. I'd recommend removing the authority logic and just using a token controlled by a simple owner, that will allow this spec to be simplified.

Reply:
The stop logic can be removed as well. Meaningless for kiln correctness.

Reply:
yeah, I was considering to use a generic gem token, maybe it could be extrapolated from DSToken removing the parts you mentioned to keep it simple and reused for future FV work as generic gem token.

Reply:
done in b9db7c7.
I've forked a `gem` erc20 token out of  `dai`, simplifying the logic and cleanup the specs as you suggested.